### PR TITLE
Fixed issue of new dag getting old dataset events.

### LINKED ISSUE #39456: DAGs are able to see historical dataset events when created new
### Apache Airflow version

2.9.1

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

If a new dataset triggreed DAG is created for an already existing dataset. (Dataset has already existing dataset events) DAG see all dataset events from very first event for dataset.

### What you think should happen instead?

If a new dataset triggreed DAG is created for an already existing dataset. (Dataset has already existing dataset events). DAG should see all the dataset events from the time when DAG was added.

### How to reproduce

* Create a new DAG which has dataset event producer. Enable both the producer.
* Run producer dag multiple times.
* Create a new consumer DAG which also get triggered from same dataset with catchup false.
* Enable the new DAG.
* You will see new DAG will run as soon as enabled and see very first event in the dataset.
* In the next run whenever that was scheduled DAG will get all the events other than the first event it has already consumed.

```from __future__ import annotations

import pendulum

from airflow.datasets import Dataset
from airflow.models.dag import DAG
from airflow.operators.bash import BashOperator

# [START dataset_def]
tosheer_dag1_dataset = Dataset("s3://tosheer-dag1/output_1.txt", extra={"partition": "bye"})


with DAG(
    dag_id="tosheer_dataset_produces_1",
    catchup=False,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule="@daily",
    tags=["produces", "dataset-scheduled"],
) as dag1:
    # [START task_outlet]
    BashOperator(outlets=[tosheer_dag1_dataset], task_id="producing_task_1", bash_command="sleep 5")
    # [END task_outlet]
```

After this add  
```
# [START dag_dep]
with DAG(
    dag_id="tosheer_dataset_consumes_1",
    catchup=False,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule=[tosheer_dag1_dataset],
    tags=["consumes", "dataset-scheduled"],
) as dag3:
    # [END dag_dep]
    BashOperator(
        task_id="consuming_1",
        bash_command='echo "ti_key={{ triggering_dataset_events }}"',
    )
```

### Operating System

Debian GNU/Linux 12 (bookworm)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Docker-Compose

### Deployment details

Plain vanilla airflow.

### Anything else?

This issue is associated with an earlier issue https://github.com/apache/airflow/issues/38826. Fix of that issue just fixed disabled - enabled / deleted - recreated.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by tosheer**
If a new dataset triggreed DAG is created for an already existing dataset. (Dataset has already existing dataset events) DAG see all dataset events from very first event for dataset.

Fixes: https://github.com/apache/airflow/issues/39456

### CODE REVIEW

File: tests/jobs/test_scheduler_job.py [hunk 1]
```diff
@@ -3776,6 +3776,11 @@ def test_create_dag_runs_datasets(self, session, dag_maker):
         dataset1 = Dataset(uri="ds1")
         dataset2 = Dataset(uri="ds2")
 
+        # Create DAG before the arrival of dataset events.
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py]: Should this be part of its own, new test?

`test_new_dagrun_ignores_old_datasets` or similar?

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -3811,20 +3816,40 @@ def test_create_dag_runs_datasets(self, session, dag_maker):
         )
         session.add(event2)
 
+        # Create a third event, creation time is more recent, but data interval is even older
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py]: Given that this is reproducible with a single, previous DAG run having a dataset event, shouldn't the test only require 3 total DAG runs?

- One DAG run from the original DAG (let's call this Run A)
- One more DAG run (Run B) after the new DAG is introduced, which triggers the new DAG's DAG run which should only have dataset info from Run B.

**Comment by jedcunningham** [comment]
cc @uranusjr

**Comment by uranusjr** [comment]
To be honest I don’t really consider the current behaviour (in 2.9.1) a bug, but I can see why people expect it this way. I think we probably need an entry in `newsfragment` to warn users about possible breakages.

**Comment by uranusjr** [comment]
Also the tests need improvements, as mentioned in above reviews.

File: tests/jobs/test_scheduler_job.py (hunk 1 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py]: @RNHTTR I've added a new unit test. Please review it and let me know if it needs any adjustments. 

Unit test is checking newly added Dag should just get the Dataset arrived after it was created(event3) and not the dataset events(event 1 and event2)

File: tests/jobs/test_scheduler_job.py (hunk 2 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py]: The newly designed test case perfectly addresses the scenario you have mentioned. Let me know if i missing anything.

**Comment by tosheer** [comment]
@RNHTTR / @uranusjr Can you guys please review it and let me know if it needs any adjustments.

File: airflow/jobs/scheduler_job_runner.py [hunk 3]
```diff
@@ -1277,6 +1277,10 @@ def _create_dag_runs_dataset_triggered(
                 ]
                 if previous_dag_run:
                     dataset_event_filters.append(DatasetEvent.timestamp > previous_dag_run.execution_date)
+                else:
+                    dataset_event_filters.append(
+                        DatasetEvent.timestamp >= DagScheduleDatasetReference.created_at
+                    )
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: Hmm just wondering if the difference between `>=` and `>` is something we should worry about.

File: tests/jobs/test_scheduler_job.py [hunk 4]
```diff
@@ -3869,13 +3870,118 @@ def dict_from_obj(obj):
...
+            timestamp=timezone.utcnow(),
+        )
+        session.add(event3)
+
+        session = dag_maker.session
+        session.add_all(
+            [
+                DatasetDagRunQueue(dataset_id=ds_id, target_dag_id=dag.dag_id),
+            ]
+        )
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        session.add(DatasetDagRunQueue(dataset_id=ds_id, target_dag_id=dag.dag_id))
```

File: tests/jobs/test_scheduler_job.py [hunk 5]
```diff
@@ -3869,13 +3870,118 @@ def dict_from_obj(obj):
...
+
+        # dag should be triggered since it only depends on dataset1, it's been queued and dataset events landed after DAG was created.
+        created_run = session.query(DagRun).filter(DagRun.dag_id == dag.dag_id).one()
+        assert created_run.state == State.QUEUED
+
+        # we don't have __eq__ defined on DatasetEvent because... given the fact that in the future
+        # we may register events from other systems, dataset_id + timestamp might not be enough PK
+        assert list(map(dict_from_obj, created_run.consumed_dataset_events)) == list(
+            map(dict_from_obj, [event3])
+        )
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        assert list(map(dict_from_obj, created_run.consumed_dataset_events)) == [dict_from_obj(event3)]
```

File: tests/jobs/test_scheduler_job.py [hunk 6]
```diff
@@ -3869,13 +3870,118 @@ def dict_from_obj(obj):
...
+        self.job_runner = SchedulerJobRunner(job=scheduler_job)
+
+        self.job_runner.processor_agent = mock.MagicMock()
+
+        with create_session() as session:
+            self.job_runner._create_dagruns_for_dags(session, session)
+
+        def dict_from_obj(obj):
+            """Get dict of column attrs from SqlAlchemy object."""
+            return {k.key: obj.__dict__.get(k) for k in obj.__mapper__.column_attrs}
```
Comment (uranusjr) [inline tests/jobs/test_scheduler_job.py:3965]:
```suggestion
            return {k.key: getattr(obj, k) for k in obj.__mapper__.column_attrs}
```

(Not sure about this)

Comment (tosheer) [reply tests/jobs/test_scheduler_job.py:3965]: Worked like a charm.

File: tests/jobs/test_scheduler_job.py (hunk 5 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py]: Done

File: tests/jobs/test_scheduler_job.py (hunk 4 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py]: Done

File: airflow/jobs/scheduler_job_runner.py (hunk 3 above)
Comment (tosheer) [reply airflow/jobs/scheduler_job_runner.py]: nope, moved it back to `>`

File: tests/jobs/test_scheduler_job.py [hunk 7]
```diff
@@ -3793,6 +3793,10 @@ def test_create_dag_runs_datasets(self, session, dag_maker):
         dataset1 = Dataset(uri="ds1")
         dataset2 = Dataset(uri="ds2")
 
+        with dag_maker(dag_id="datasets-consumer-single", schedule=[dataset1]):
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py:3796]: Why is this block being moved?

File: tests/jobs/test_scheduler_job.py [hunk 8]
```diff
@@ -3830,9 +3834,6 @@ def test_create_dag_runs_datasets(self, session, dag_maker):
 
         with dag_maker(dag_id="datasets-consumer-multiple", schedule=[dataset1, dataset2]):
             pass
-        dag2 = dag_maker.dag
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py:3833]: Is this being removed for a reason?

File: tests/jobs/test_scheduler_job.py [hunk 9]
```diff
@@ -3869,13 +3870,112 @@ def dict_from_obj(obj):
...
-        assert session.query(DatasetDagRunQueue).filter_by(target_dag_id=dag2.dag_id).one() is not None
+        assert session.query(DatasetDagRunQueue).filter_by(target_dag_id=dag3.dag_id).one() is not None
         # dag2 should not be triggered since it depends on both dataset 1  and 2
-        assert session.query(DagRun).filter(DagRun.dag_id == dag2.dag_id).one_or_none() is None
+        assert session.query(DagRun).filter(DagRun.dag_id == dag3.dag_id).one_or_none() is None
         # dag3 DDRQ record should be deleted since the dag run was triggered
-        assert session.query(DatasetDagRunQueue).filter_by(target_dag_id=dag3.dag_id).one_or_none() is None
+        assert session.query(DatasetDagRunQueue).filter_by(target_dag_id=dag2.dag_id).one_or_none() is None
+
+        assert dag2.get_last_dagrun().creating_job_id == scheduler_job.id
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py:3879]: Should there be any changes to this test (`test_create_dag_runs_datasets`) at all?

File: tests/jobs/test_scheduler_job.py [hunk 10]
```diff
@@ -3869,13 +3870,112 @@ def dict_from_obj(obj):
...
+            source_dag_id=dr.dag_id,
+            source_run_id=dr.run_id,
+            source_map_index=-1,
+        )
+        session.add(event2)
+
+        # Create DAG after dataset events.
+        with dag_maker(dag_id="datasets-consumer", schedule=[dataset]):
+            pass
+        dag = dag_maker.dag
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py]: Can you make this and following references to `dag` be `consumer_dag` just to make it more obvious?

File: tests/jobs/test_scheduler_job.py [hunk 11]
```diff
@@ -3869,13 +3870,112 @@ def dict_from_obj(obj):
...
+        def dict_from_obj(obj):
+            """Get dict of column attrs from SqlAlchemy object."""
+            return {k.key: getattr(obj, k) for k in obj.__mapper__.column_attrs}
+
+        # dag should be triggered since it only depends on dataset1, it's been queued and dataset events landed after DAG was created.
+        created_run = session.query(DagRun).filter(DagRun.dag_id == dag.dag_id).one()
+        assert created_run.state == State.QUEUED
+
+        # we don't have __eq__ defined on DatasetEvent because... given the fact that in the future
+        # we may register events from other systems, dataset_id + timestamp might not be enough PK
```
Comment (RNHTTR) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        # __eq__ isn't defined on DatasetEvent
```
IMO this is sufficient

File: tests/jobs/test_scheduler_job.py (hunk 7 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py:3796]: @RNHTTR The DAG needs to be created first before the event can arrive for the DAG to see all the events.

File: tests/jobs/test_scheduler_job.py (hunk 9 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py:3879]: yes, as now any dag which was created before the event can arrive will see only the event and i want to keep the dag created first named dag2 and dag created later named dag3

File: tests/jobs/test_scheduler_job.py (hunk 10 above)
Comment (tosheer) [reply tests/jobs/test_scheduler_job.py]: That makes sense. Update is done.

**Review by uranusjr — APPROVED** [review]
Looks good to me. Pending @RNHTTR’s comments.

**Review by RNHTTR — APPROVED** [review]

**Comment by RNHTTR** [comment]
> Looks good to me. Pending @RNHTTR’s comments.

LGTM as well, but I'm unable to mark the comments as resolved :\

**Comment by tosheer** [comment]
@RNHTTR All tests are passing now. 

@uranusjr can you merge the PR?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Tests are failing.

**Comment by tosheer** [comment]
@potiuk Test are passing now. Can you merged
Screenshot 2024-06-19 at 1 12 50 PM

**Comment by Dev-iL** [comment]
(Lesson for myself - don't post PR comments on Slack. Anyway, here's a copy of [my comment from a few days ago](https://apache-airflow.slack.com/archives/CCZRF2U5A/p1718282492028369?thread_ts=1715668779.756509&cid=CCZRF2U5A):)

Perhaps it makes sense to have this behaviour user-configurable? Seeing past dataset events sounds useful for backfilling (i.e. when a new process becomes available for old data).

It's not clear from the docs how [`catchup`](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dag-run.html#catchup) works for dataset schedules, but I'd imagine that:
- dataset schedule + `catchup=True` -> current behavior (see all past events]).
- dataset schedule + `catchup=False` -> the behavior you added (see only new events).

If we take the route suggested above, we should consider what it'd mean for a `DatasetOrTimeSchedule` + `catchup=True` (I'd say either forbidden or add all runs from both schedules).

File: tests/jobs/test_scheduler_job.py (hunk 6 above)
Comment (Dev-iL) [reply tests/jobs/test_scheduler_job.py:3965]: I believe `vars(obj)` is preferred to `obj.__dict__`

**Comment by uranusjr** [comment]
3.0 is a point when we can break things so let’s consider this before we do that.

**Pull request closed without merging**